Support multi-deck cards #495, and multiple decks in a single note #705

ronzulu opened the pull request:
This PR addresses issues:
#178 
#495
#705 

For example, in the following:
```
#flashcards/language/words #flashcards/trivia #flashcards/learned-from-tv

A group of cats is called a::clowder

#flashcards/language/words
A group of crows is called a::murder
The branch of medical science dealing with the classification of diseases::nosology

#flashcards/trivia 
The 1988 movie "Mystic Pizza" launched the career of what 'Pretty Woman?' actress::Julia Roberts
What is Australia’s most popular spread?::Vegemite
Who was Australia’s first female prime minister?::Julia Gillard
```

| Question | Decks | 
|-|-|
| A group of cats is called a::clowder | #flashcards/language/words #flashcards/trivia #flashcards/learned-from-tv | 
| A group of crows is called a::murder | #flashcards/language/words | 
| The branch of medical science dealing with the classification of diseases::nosology | #flashcards/language/words | 
| The 1988 movie "Mystic Pizza" launched the career of what 'Pretty Woman?' actress::Julia Roberts | #flashcards/trivia  | 
Etc

# Syntax

For a question to appear in multiple decks, preceding the question in the file must be the list of tags on the same line, such as:
```
#flashcards/language/words #flashcards/trivia #flashcards/learned-from-tv
A group of cats is called a::clowder
```

The list of deck(s) can be changed using this format multiple times within a single note.

The list of deck(s) can be included in the frontmatter, in which case multiple tags can be on multiple lines, such as:
```
---
tags:
  - flashcards
  - flashcards/learned-from-tv
---
```

A question specific tag takes precedence over the note-level tags described above. For example, in the following:
- "Q1::A1" is in deck `#flashcards/spatial`
- "Which 1986 movie..." is in deck  `#flashcards/trivia`, i.e. the last note-level tag

```
#flashcards/trivia 
The 1988 movie "Mystic Pizza" launched the career of what 'Pretty Woman?' actress::Julia Roberts

#flashcards/spatial Q1::A1

Which 1986 movie is the highest-grossing Australian movie of all time?::Crocodile Dundee
```

# Card counting

Multi-deck cards are only counted once, e.g.:
- There are 6 cards, so at (1) the total shown is 6

In the case of multi-deck cards, totals may not add up. E.g. under `flashcards` is `language`, `learned-from-tv` and `trivia`. Those decks have 3, 1, 4 cards listed respectively, totaling 8. Which is 2 more than the total of 6 shown at (1). This is because the `A group of cats is called a::clowder` question is present in 3 decks, but is only counted as if it was in one.

![image](https://github.com/st3v3nmw/obsidian-spaced-repetition/assets/75528127/883ad821-c4b9-4e08-967b-31b60c0a719b)

# Reviewing

A multi-deck card will only be shown once. For example, after reviewing only the `learned-from-tv` deck, the counts are:
![image](https://github.com/st3v3nmw/obsidian-spaced-repetition/assets/75528127/695dc18e-49de-4e94-9f08-f6734b59d919)

The new-card counts in blue, across the 3 decks in which the cat question appears, have all decreased by 1.

i.e. lines (3) (4) and (5). (2) has also decreased as it is the parent of (3).

# Backward Compatibility

For compatibility with plug-in functionality before this PR, if there are no note-level tags prior to a question, and that question doesn't have a question specific tag, the question will be placed in the deck specified by the first tag found after the question within the file.

e.g. if the complete content of a note is:
```
The 1988 movie "Mystic Pizza" launched the career of what 'Pretty Woman?' actress::Julia Roberts
What is Australia’s most popular spread?::Vegemite
Who was Australia’s first female prime minister?::Julia Gillard

#flashcards/backward-compatibility
```

All the questions will be found in deck `#flashcards/backward-compatibility`

# Beta Testing & Feedback

Yes please!

ronzulu commented:
Hi all

Beta release is now available.

There is a file called `main.js` in this zip file
[main_495A.zip](https://github.com/st3v3nmw/obsidian-spaced-repetition/files/13909934/main_495A.zip)

You will need to copy that to the obsidian plug-in folder, e.g.
`D:\Obsidian\Obsidian\.obsidian\plugins\obsidian-spaced-repetition`

Then reload the plug-in, or restart obsidian. 

Any feedback appreciated.

Ronny

pikatwinky commented:
Hi Ronny, thank you so much for your hard work. All we use this plugin really appreciate it.

Im not very tech skilled, but i wanted to give a try to this beta version (using the sandbox vault first and then with my own vault also restarting the plugin in the second case); so i copied the main.js file in the plugin route, but when using it my existent cards are not shown in the review window
![image](https://github.com/st3v3nmw/obsidian-spaced-repetition/assets/31860235/253acd79-e686-4bfc-9de9-645c86d67235)

ronzulu commented:
Hi @pikatwinky thanks for trying, and oops that's not a good start!

Please attach one of your Markdown .md files so I can isolate the problem.

Also, can you send a screenshot of the relevant settings, including:
![image](https://github.com/st3v3nmw/obsidian-spaced-repetition/assets/75528127/6b6b1191-158d-4248-982a-29a4441bc336)

Cheers
Ronny

Steindvart commented:
A great addition! This is exactly what I REALLY miss in using this plugin 😅.

In basic case is work fine in my test Obsidian vault.

![image](https://github.com/st3v3nmw/obsidian-spaced-repetition/assets/53004350/af0c0652-137e-44cd-9148-ea1c182ab7aa)

![image](https://github.com/st3v3nmw/obsidian-spaced-repetition/assets/53004350/454393c5-3af8-4d4b-af62-5c09b1829a38)

![image](https://github.com/st3v3nmw/obsidian-spaced-repetition/assets/53004350/a3f13bba-4e1d-4fbc-831d-d13c1038a45c)

pikatwinky commented:
Hi @ronzulu, i tried with the sandbox vault and with my own vault. In both cases i use this settings
![image](https://github.com/st3v3nmw/obsidian-spaced-repetition/assets/31860235/8e5f9ad3-aeda-49d7-9aba-053a9cdbb8a5)

The md i use in each case looks like this
Sandbox vault
![image](https://github.com/st3v3nmw/obsidian-spaced-repetition/assets/31860235/3f4eda12-13b8-40eb-9ada-1faad440c2e7)
My own vault:
![image](https://github.com/st3v3nmw/obsidian-spaced-repetition/assets/31860235/e6226287-8ca9-4898-a823-a92d2f82179c)

I copied the main.js file in the plugin's route, but at reviewing time the cards aren't shown and the window looks as in my previous attachment
I tried restarting both the plugin and obsidian. Im using the latest obsidian version btw

Lucas256-cmd commented:
So I had a file with a lot of hierarchical (if this is even a word:) ) card tags. 3 layers deep, some with one deck in them some with more. I downloaded the zip, put the file where it belongs, reloaded obsidian AND IT WORKED LIKE A CHARM. Thank you. This is exactly how i had it in mind since the tags where already in the file before you added this functionality. You added it and it just works.

ronzulu commented:
Hi @pikatwinky  from your screenshots you have done everything correctly.

Could you please zip up your test vault and post here. Hopefully I'll be able to reproduce the issue at my end.

Thanks
Ronny

Lucas256-cmd commented on CHANGELOG.md:
No problem

ronzulu replied:
Hi @Steindvart I'm quite new to TypeScript and open source projects on GitHub, so any feedback is appreciated. 

I didn't quite understand your point. Do you mean that after I edit the `changelog.md` file, the EOL character becomes `\r\n`?

I'm using Visual Studio Code running under Windows, and Windows uses `\r\n` as the EOL character.

What issue do you see at your end?

Spasibo (one of the few words in Russian I know)

Steindvart replied:
@ronzulu,

I ran into a similar problem with this unexpected changes. As I understand it, this happens after executing the `pnpm format`. Code style format, apparently, it also affects other text files besides the source code files. And adds a line break `\n` to the end if it is not there. So we get such "unexpected" changes.

For example. Binary structure of file, before `format`:
![00](https://github.com/st3v3nmw/obsidian-spaced-repetition/assets/53004350/f298757a-2501-46e9-899f-343c24a3d411)

And after:
![0](https://github.com/st3v3nmw/obsidian-spaced-repetition/assets/53004350/8e044e70-7f35-4794-b221-88d053061a1c)

Regarding `\n` and `\r\n` - such substitutions can be made by the system or text editor itself. Therefore, I initially assumed this hypothesis. But, as I wrote above, the reason is the code style formatter.

Steindvart replied:
@ronzulu

> Spasibo (one of the few words in Russian I know)

My pleasure :) Thank you for such a sensitive and kind attitude. I appreciate it.

ronzulu commented:
> Hi @pikatwinky from your screenshots you have done everything correctly.
> 
> Could you please zip up your test vault and post here. Hopefully I'll be able to reproduce the issue at my end.
> 
> Thanks Ronny

Sorry to hassle you, would be great if you can send this through. Many thanks!

pikatwinky commented:
> > Hi @pikatwinky from your screenshots you have done everything correctly.
> > Could you please zip up your test vault and post here. Hopefully I'll be able to reproduce the issue at my end.
> > Thanks Ronny
> 
> Sorry to hassle you, would be great if you can send this through. Many thanks!

Hi Ronny, sorry for the late reply. I had that problem using the sandbox vault and a personal vault. Unfortunately i cant publish the personal one. Does the sandbox work for you?
Maybe i did something wrong with my process which was the next one:
-Open Sandbox vault
-Download Spaced Repetition plugin
-Open Obsidian's Spaced Repetition plugin folder and paste there the beta main.js file
-When returning Obsidian plugin wasn't working even after restarting from the plugins section

ronzulu commented:
Hi @pikatwinky 

I don't know why I didn't think of it before, but here is a demo vault with the beta version of `main.js`.
[Vault_495A.zip](https://github.com/st3v3nmw/obsidian-spaced-repetition/files/14004789/Vault_495A.zip)

When I open the vault, and click on the flashcards icon, I see:
![image](https://github.com/st3v3nmw/obsidian-spaced-repetition/assets/75528127/b2680ed7-b538-4165-83ff-68f8d3795888)

What do you see?

If you see the same as above, perhaps copy one of your test files into this vault and see what happens. If the beta code doesn't work with your file in this demo vault, please zip it up and post the zip file here.

Thanks for your help!

Cheers
Ronny

pikatwinky commented:
> Hi @pikatwinky
> 
> I don't know why I didn't think of it before, but here is a demo vault with the beta version of `main.js`. [Vault_495A.zip](https://github.com/st3v3nmw/obsidian-spaced-repetition/files/14004789/Vault_495A.zip)
> 
> When I open the vault, and click on the flashcards icon, I see: ![image](https://private-user-images.githubusercontent.com/75528127/298460907-b2680ed7-b538-4165-83ff-68f8d3795888.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU5NDE1MDUsIm5iZiI6MTcwNTk0MTIwNSwicGF0aCI6Ii83NTUyODEyNy8yOTg0NjA5MDctYjI2ODBlZDctYjUzOC00MTY1LTgzZmYtNjhmOGQzNzk1ODg4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTIyVDE2MzMyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI3YTdiNDM4NjVhZWFhMDIzN2FhNjBlY2IxZGJmNGJhZjk2NmRjYmI0YTBiOGZiOGRhZGQ1OWYyNzc1OGMyMTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.0AT7GfWNJhC49xk4Kec2iLPfoj_0rIuEnfqwShcltW8)
> 
> What do you see?
> 
> If you see the same as above, perhaps copy one of your test files into this vault and see what happens. If the beta code doesn't work with your file in this demo vault, please zip it up and post the zip file here.
> 
> Thanks for your help!
> 
> Cheers Ronny

Hi @ronzulu.
I tested your vault and it works perfectly and the behaviour is as described.
Thank you so much for your hard work!

ronzulu commented:
Hi @pikatwinky great.

Does that mean it works when you copy your file into the test vault?

I assume that the problem still occurs in your main vault? If so, I'll need to send you a version with extra debug to assist diagnosing.

Cheers
Ronny

pikatwinky commented:
> > > Hi @pikatwinky from your screenshots you have done everything correctly.
> > > Could you please zip up your test vault and post here. Hopefully I'll be able to reproduce the issue at my end.
> > > Thanks Ronny
> > 
> > 
> > Sorry to hassle you, would be great if you can send this through. Many thanks!
> 
> Hi Ronny, sorry for the late reply. I had that problem using the sandbox vault and a personal vault. Unfortunately i cant publish the personal one. Does the sandbox work for you? Maybe i did something wrong with my process which was the next one: -Open Sandbox vault -Download Spaced Repetition plugin -Open Obsidian's Spaced Repetition plugin folder and paste there the beta main.js file -When returning Obsidian plugin wasn't working even after restarting from the plugins section

Hi @ronzulu! If the steps i followed here are all right, then yes, all my existent cards still dissapear from the review window, but the other plugin features like the note review work properly.
How can i provide you info about this?

ronzulu commented:
Hi @pikatwinky 

No worries, I understand you can't send through your personal vault. I'm hoping to be able to reproduce the problem on my machine to simplify solving it.

1. Is there one file from your vault that contains cards that you can copy to the test vault that I sent? Let's call it `file 1`

2. What happens in the test vault?

When you click on the flashcards icon, do the cards from `file 1` appear, or only the ones that were originally present in the test vault?
![image](https://github.com/st3v3nmw/obsidian-spaced-repetition/assets/75528127/b2680ed7-b538-4165-83ff-68f8d3795888)

3. Please take a screenshot of the flashcard review dialog box.

4. In file 1, are you able to make this as compact as possible that still highlights the problem? E.g. remove all unnecessary text and just keep a single card.

Replace any private info in the file with just boring general text, retaining the flashcard tags.

5. Does the problem still occur?

6. As this doesn't have any private info, do you feel comfortable zipping up this test vault and sending it to me?

Sorry for all this hassle, but I think this is a useful new feature and there isn't any point releasing it if it doesn't work for you!

Cheers
Ronny

pikatwinky commented:
Hi @ronzulu 
Here i share with you 3 vaults testing the next simple note:
![image](https://github.com/st3v3nmw/obsidian-spaced-repetition/assets/31860235/21fc8392-9c38-4687-9152-1189f9444e9f)
I'm using the default settings
- The first is a test vault with the plugin's latest release and one note with one flashcard. In this vault the review window works well 
![image](https://github.com/st3v3nmw/obsidian-spaced-repetition/assets/31860235/67a46a70-3653-49a5-8a7f-f028465e4b79)
[MyTestVault.zip](https://github.com/st3v3nmw/obsidian-spaced-repetition/files/14073781/MyTestVault.zip)

- Then in the second vault i copied the beta main.js attached above in the corresponding plugin's route, and after restarting obsidian, the review window doesn't show the card 
![image](https://github.com/st3v3nmw/obsidian-spaced-repetition/assets/31860235/ff9c2025-8973-44ad-ac45-31355486e0a4)
[MyTestVaultAfterUpdate.zip](https://github.com/st3v3nmw/obsidian-spaced-repetition/files/14073782/MyTestVaultAfterUpdate.zip)

- Finally in the vault you shared previously with the beta plugin included, i put this same note but i changed its deck's tag to **flashcardsTest** only for easy identification.  
![image](https://github.com/st3v3nmw/obsidian-spaced-repetition/assets/31860235/755d2f50-89d9-4679-a2ad-4aaf103bbcc4)
However the flashcard doesn't appear in the review window
![image](https://github.com/st3v3nmw/obsidian-spaced-repetition/assets/31860235/f9a9eded-a299-4f06-ab6f-28a35cd6871c)

I was expecting a **flashcardsTest** deck appeared in the window because that tag was added in the settings
![image](https://github.com/st3v3nmw/obsidian-spaced-repetition/assets/31860235/5855edc5-4acf-45bf-adb3-5cb0d579d395)

[Demo.zip](https://github.com/st3v3nmw/obsidian-spaced-repetition/files/14073817/Demo.zip)

I hope this info could be found useful

ronzulu commented:
Hi @pikatwinky 

Thanks so much for the clear and comprehensive info, it enabled me (hopefully) to solve the problem.
![image](https://github.com/st3v3nmw/obsidian-spaced-repetition/assets/75528127/93b1f8e7-d173-40bc-b88a-239dd2b86364)

![image](https://github.com/st3v3nmw/obsidian-spaced-repetition/assets/75528127/18df57e3-fe95-49b1-83cb-154a0f264693)


If you could please try this version now, and let me know how it goes.
[main_495B.zip](https://github.com/st3v3nmw/obsidian-spaced-repetition/files/14075882/main_495B.zip)

Cheers
Ronny

pikatwinky commented:
> Hi @pikatwinky
> 
> Thanks so much for the clear and comprehensive info, it enabled me (hopefully) to solve the problem. ![image](https://private-user-images.githubusercontent.com/75528127/300259866-93b1f8e7-d173-40bc-b88a-239dd2b86364.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY1ODMyNDksIm5iZiI6MTcwNjU4Mjk0OSwicGF0aCI6Ii83NTUyODEyNy8zMDAyNTk4NjYtOTNiMWY4ZTctZDE3My00MGJjLWI4OGEtMjM5ZGQyYjg2MzY0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTMwVDAyNDkwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNlZWIxYjE4NTUxMDY2NmUxZTBhNTBhYTc3NmY3M2NjNjM4OGQ1YmM3ODRlMGFjMTkzZTU0OWRjNjVlYWZjNTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.k9oe_oYKNgO0j4FGsPjk0IyTX28FgXH8-XVbEVLoozc)
> 
> ![image](https://private-user-images.githubusercontent.com/75528127/300259967-18df57e3-fe95-49b1-83cb-154a0f264693.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY1ODMyNDksIm5iZiI6MTcwNjU4Mjk0OSwicGF0aCI6Ii83NTUyODEyNy8zMDAyNTk5NjctMThkZjU3ZTMtZmU5NS00OWIxLTgzY2ItMTU0YTBmMjY0NjkzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTMwVDAyNDkwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ1N2UyNDk0NWNjODlmM2M2ZjExMmMxNzU0YmQyY2UxMWZhNDA2OWMzNzc5N2I5YzdmNTA4ZjU5YjRmNTVkZGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.uxwqJnPzRAksRiFt1h7PYtMmTVP87k1tcPwhKHX3nJ8)
> 
> If you could please try this version now, and let me know how it goes. [main_495B.zip](https://github.com/st3v3nmw/obsidian-spaced-repetition/files/14075882/main_495B.zip)
> 
> Cheers Ronny

Hi @ronzulu, i tested this new version in my own vault and it works perfectly, thanks again for all your hard work
![image](https://github.com/st3v3nmw/obsidian-spaced-repetition/assets/31860235/e448e605-3e85-4344-acbc-d58adf15f1e0)

ronzulu commented:
Hi @st3v3nmw this has been tried out by a few people, and I've fixed the one bug reported.

This PR has required moderate change (more than I anticipated), so a review by you really is needed. The original PR request has a comprehensive description of the enhanced functionality, and I have also updated the English user documentation to match.

Also, will you have time at some stage for us to talk through all the PRs contributed and create a plan? 

Cheers
Ronny

ronzulu commented:
Hi @st3v3nmw,  just wondering how you are - you must be very busy 😄 

It would be great if you could review and merge this PR.

> Hi @st3v3nmw this has been tried out by a few people, and I've fixed the one bug reported.
> 
> This PR has required moderate change (more than I anticipated), so a review by you really is needed. The original PR request has a comprehensive description of the enhanced functionality, and I have also updated the English user documentation to match.
> 
> Also, will you have time at some stage for us to talk through all the PRs contributed and create a plan?
> 
> Cheers Ronny

st3v3nmw reviewed (approved):
Hi @ronzulu,

I had tested this manually and it looks good to me. Thanks!
